Time regions support in graph panel

marefr opened the pull request:
Fixes #5930 

Example dashboard using Grafana Testdata datasource: https://gist.github.com/marefr/2fee183a7b566ab12c6477fccec5a887

![image](https://user-images.githubusercontent.com/1668778/44036615-7b44cbe2-9f12-11e8-99ed-1f16f14705d3.png)

![image](https://user-images.githubusercontent.com/1668778/44036625-812f13f0-9f12-11e8-83c2-7874a7329d80.png)

![image](https://user-images.githubusercontent.com/1668778/44036636-86913d00-9f12-11e8-8dbd-eecb8d732fa9.png)

![image](https://user-images.githubusercontent.com/1668778/44036643-8a29f218-9f12-11e8-9b0e-d043f01e61ca.png)

marefr committed:
graph: Time region support

torkelo commented:
I wonder how much conflict this will have with the major refactorings in https://github.com/grafana/grafana/pull/12060/files

marefr commented:
Good question :)

torkelo commented:
Maybe thresholds are the wrong name for this? Been thinking about thresholds recently and how we can align the thresholds model  & UI across panels. But this feels a lot different. more like `Colored Time Regions` or Time Markings.

marefr commented:
Yes agree, but you mean thresholds in general? Colored Time Regions or Time Markings are exactly the same as Colored Value Regions or Value Markings, right?

marefr commented:
@torkelo did some work on this:
* No longer combined with thresholds and moved to a separate section under Display tab
* The whole concept/feature is now called "time regions"

perplexityjeff commented:
This is a great feature that we are looking for to be added to Grafana. Is there any update for this request at this moment?

marefr commented:
@perplexityjeff hopefully it can be included in v5.4.

marefr committed:
devenv: graph time regions test dashboard

marefr committed:
create time regions solely based on utc time

marefr committed:
devenv: graph time regions test dashboard

marefr committed:
docs: description about graph panel time regions feature

StephenOTT commented:
Any plans to add this into other panel types such as HeatMap? thinking prometheus histogram heatmap usage as a example.

marefr committed:
set default color mode

marefr committed:
fix time regions mutable bug

marefr committed:
devenv: graph time regions test dashboard

marefr commented:
@StephenOTT not at the moment. That panel is using d3 so it's not that easy to implement just by copy this solution. Please open a separate issue for heatmap time regions.

torkelo commented on docs/sources/features/panels/graph.md:
```diff
@@ -186,6 +186,14 @@ There is an option under Series overrides to draw lines as dashes. Set Dashes to
+
+> Only available in Grafana v5.4 and above.
+
+{{< docs-imagebox img="/img/docs/v54/graph_time_regions.png" max-width= "800px" >}}
```
Is this image in the grafana.org repo?

marefr replied:
Yes in description

> PR for docs screenshot: grafana/grafana.org#77

marefr replied:
Will push that image to master directly now.

torkelo reviewed (approved):
Looks really great. Sorry for taking for every to review & test this. 

I would remove the background %3 & %6 and rename background %9 with Gray. And maybe add a blue transparent color as well.

marefr committed:
changed time region color modes

marefr commented:
Replaced background color modes with Gray. Added blue color. Moved custom to bottom.

![image](https://user-images.githubusercontent.com/1668778/48565589-321c0d00-e8f9-11e8-8f3d-f9b8b4d8ec12.png)

marefr committed:
Add tooltip

Gauravshah commented:
Are there plans to support timezone ? if there are some directions we would love to contribute

marefr commented:
@Gauravshah what do you mean? You configure times that correlates to UTC. Changing timezone of dashboard to local browser time will adjust the time regions accordingly.

Gauravshah commented:
@marefr lets say my timezone is PT ( Los Angeles ). I select `Sat` as my time region.  It would show on my graph as Friday night 8 PM to Saturday 8 PM. Then lets say then I pick `Sat 8 AM` to `Sun 8:AM` which will correctly show as Saturday PT. But this would go off when we change DST. We would need to go back and fix all dashboards from 8 AM to 7 AM

marefr commented:
I would say then use UTC time in dashboard as timezone instead or open a feature request issue.

Gauravshah commented:
@marefr I think there are valid uses cases around using time zone. Will open a feature request.
